Use the new topics endpoint in Connect REST API to describe the topics of a connector

kkonstantine opened the pull request:
### Description 
After KIP-558 and starting with Apache Kafka version 2.5, Kafka Connect adds a new 'topics' endpoint per connector to discover the set of topics a connector is using during runtime. Using this new endpoint will allow ksqlDB to get a more accurate view of a connector's active topics. The code is refactor to no longer depend on listing the topics with an adminclient when describing a connector in ksqlDB

Details on active topics tracking here: 
https://cwiki.apache.org/confluence/display/KAFKA/KIP-558%3A+Track+the+set+of+actively+used+topics+by+connectors+in+Kafka+Connect

### Testing done 
Existing unit tests were adjusted to test the new functionality. 

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

agavra commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutor.java:
```diff
@@ -114,16 +115,23 @@ public DescribeConnectorExecutor() {
+        .topics(connectorName);
+
+    final List<KsqlWarning> warnings;
+    final List<String> topics;
```
we should use the information here before getting the data sources above (basically instead of calling `matches` on L106 above we should check `topics#contains`)

kkonstantine replied:
Moved up with a small optimization to avoid hitting the `topics` endpoint if the first REST call did not return the connector's info. But if the doubly nested if does look great or you think it's worth calling regardless I can change the if/else

kkonstantine commented on ksql-engine/src/main/java/io/confluent/ksql/services/DefaultConnectClient.java:
```diff
@@ -214,6 +224,33 @@ public DefaultConnectClient(
   }
 
+  @Override
+  public ConnectResponse<Map<String, Map<String, List<String>>>> topics(final String connector) {
```
The reason is that the key is the actual name of the connector and not `connector : <connector name>`

Maybe my json foo is missing a way to deserialize this with a `JsonCreator` and the constructor, but based on what I know it can't be done with jackson and the return type that you mention (which is a bummer but we use that notation in other endpoints already). The return type is a bit verbose, indeed.

kkonstantine commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutor.java:
```diff
@@ -46,6 +45,8 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(DescribeConnectorExecutor.class);
+  @VisibleForTesting
+  protected static final String TOPICS_KEY = "topics";
```
I don't us package private that much for members that are meant to be visible in tests, but happy to adjust if that's the default in the project.

kkonstantine commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/DescribeConnectorExecutor.java:
```diff
@@ -137,16 +145,4 @@ public DescribeConnectorExecutor() {
-
-    return admin.listTopics().names().get(5, TimeUnit.SECONDS)
-        .stream()
-        .filter(connector.get()::matches)
```
Removed. Seems that there's no need for `Connector#matches` and `Connector#mapToSource` as well as their respect member fields. I removed a couple unit tests that were built around those two methods and will wait for a full build to see what else might need adjustment (I have some issues running the full test suite locally).

kkonstantine commented:
Thanks for the review @agavra !
I have addressed your comments, let me know if the removal of `topic-prefix` looks good. 
I had to rebase to get some recent changes.
Waiting for a green build.
